File: app/src/main/java/com/kimjisub/launchpad/manage/Log.java
package com.kimjisub.launchpad.manage;

public class Log {
	public static final boolean DEBUG = true;

	public static void log(String msg) {
		printLog("com.kimjisub._log", msg);
	}

	public static void test(String msg) {
		printLog("com.kimjisub._test", msg);
	}

	public static void network(String msg) {
		printLog("com.kimjisub._network", msg);
	}

	public static void activity(String msg) {
		printLog("com.kimjisub._activity", msg);
	}

	public static void firebase(String msg) {
		printLog("com.kimjisub._firebase", msg);
	}

	public static void vungle(String msg) {
		printLog("com.kimjisub._vungle", msg);
	}

	public static void admob(String msg) {
		printLog("com.kimjisub._admob", msg);
	}

	public static void midi(String msg) {
		printLog("com.kimjisub._midi", msg);
	}

	public static void midiDetail(String msg) {
		printLog("com.kimjisub._mididetail", msg);
	}

	public static void driver(String msg) {
		printLog("com.kimjisub._driver", msg);
	}

	public static void sqlite(String msg) {
		printLog("com.kimjisub._sqlite", msg);
	}

	public static void driverCycle(String msg) {
		printLog("com.kimjisub._cycle", msg);
	}

	public static void err(String msg) {
		printLog("com.kimjisub._err", msg);
	}

	private static void printLog(String tag, String msg) {

		if (msg == null)
			msg = "(null)";

		String space = "";
		for (int i = 0; i < 30 - tag.length(); i++)
			space += " ";

		String trace1 = trace(Thread.currentThread().getStackTrace(), 5);
		String trace2 = trace(Thread.currentThread().getStackTrace(), 4);
		String detailMsg = String.format("%s%-50s%-40s%-40s", space, msg, trace1, trace2);

		android.util.Log.e(tag, DEBUG ? detailMsg : msg);
	}

	public static String trace(final StackTraceElement e[], final int level) {
		if (e != null && e.length >= level) {
			final StackTraceElement s = e[level];

			if (s != null)
				return e[level].getMethodName() + "(" + e[level].getFileName() + ":" + e[level].getLineNumber() + ")";
		}
		return null;
	}
}

File: manager/src/main/java/com/kimjisub/manager/Log.kt
package com.kimjisub.manager

import android.util.Log

object Log {
	private const val DEBUG = true
	fun log(msg: String) = printLog("com.kimjisub._log", msg)
	fun fbmsg(msg: String) = printLog("com.kimjisub._fbmsg", msg)
	fun test(msg: String) = printLog("com.kimjisub._test", msg)
	fun download(msg: String) = printLog("com.kimjisub._download", msg)
	fun network(msg: String) = printLog("com.kimjisub._network", msg)
	fun activity(msg: String) = printLog("com.kimjisub._activity", msg)
	fun firebase(msg: String) = printLog("com.kimjisub._firebase", msg)
	fun admob(msg: String) = printLog("com.kimjisub._admob", msg)
	fun midi(msg: String) = printLog("com.kimjisub._midi", msg)
	fun thread(msg: String) = printLog("com.kimjisub._thread", msg)
	fun midiDetail(msg: String) = printLog("com.kimjisub._mididetail", msg)
	fun driver(msg: String) = printLog("com.kimjisub._driver", msg)
	fun driverCycle(msg: String) = printLog("com.kimjisub._cycle", msg)
	fun err(msg: String) = printLog("com.kimjisub._err", msg)

	private fun printLog(tag: String, msg: String?) {
		val msg: String = msg ?: "(null)"
		var space = ""
		for (i in 0 until 30 - tag.length) space += " "
		val trace1 = trace(Thread.currentThread().stackTrace, 5)
		val trace2 = trace(Thread.currentThread().stackTrace, 4)
		val detailMsg = String.format("%s%-50s  %-40s  %-40s", space, msg, trace1, trace2)
		Log.e(tag, if (DEBUG) detailMsg else msg)
	}

	fun trace(e: Array<StackTraceElement>?, level: Int): String? {
		if (e != null && e.size >= level)
			return "${e[level].methodName}(${e[level].fileName}:${e[level].lineNumber})"
		return null
	}
}